### PR TITLE
Add support for SMTP authentication: AUTH PLAIN.

### LINKED ISSUE #14634: AUTH PLAIN only smtp servers cannot be used
This is a (multiple allowed):

* [x] bug
* [x] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 3.8.6
* Platform and Target: doesnt matter

### What you did
Use the framework to send mails via smtp on server that only allow AUTH PLAIN authentification

### What happened
`535 5.7.8 Error: authentication failed: Invalid authentication mechanism`

### What you expected to happen
A mail should have been sent.

Possible duplicate: #13708 

There are multiple providers out there that only allow AUTH PLAIN. It should be supported, there is no reason why this is limited.

Relevant code: https://api.cakephp.org/3.8/source-class-Cake.Mailer.Transport.SmtpTransport.html#252-279

### PR DESCRIPTION
**Pull request by ADmad**
Fixes #14634

### CODE REVIEW

File: src/Mailer/Transport/SmtpTransport.php [hunk 1]
```diff
@@ -271,26 +271,42 @@ protected function _connect(): void
...
+
+        $username = $this->_config['username'];
+        $password = $this->_config['password'];
+
+        $replyCode = $this->_smtpSend(
+            sprintf(
+                'AUTH PLAIN %s',
+                base64_encode($username . chr(0) . $username . chr(0) . $password)
+            ),
+            '235|504|535'
```
Comment (ADmad) [inline src/Mailer/Transport/SmtpTransport.php:286]: Here I am checking for `504` and `535` to allow falling back to `AUTH LOGIN`. 

garas suggested checking `504 5.7.4 Unrecognized Authentication Type` but the ticket author mentioned his server returned `535 5.7.8 Error: authentication failed: Invalid authentication mechanism` when using `AUTH LOGIN` when it only supports `AUTH PLAIN`. So I am guess if the opposite was true then too a server could return `535` and not `504`.

**Comment by ADmad** [comment]
`Static Analysis` is unrelated to these changes and addressed in another PR.

File: src/Mailer/Transport/SmtpTransport.php (hunk 1 above)
Comment (garas) [reply src/Mailer/Transport/SmtpTransport.php:286]: [RFC 4954 - SMTP Service Extension for Authentication](https://tools.ietf.org/html/rfc4954#section-6):

- 504 5.7.4 Unrecognized authentication type
- 534 5.7.9  Authentication mechanism is too weak
- 535 5.7.8  Authentication credentials invalid

It looks server should reply with 534, but replied with 535.

With wrong credentials, it should stop here and not retry `AUTH LOGIN`. But if there is non-compliant servers, then maybe retry if server responded `504|534|535`

File: src/Mailer/Transport/SmtpTransport.php [hunk 2]
```diff
@@ -271,26 +271,42 @@ protected function _connect(): void
...
+            return;
+        }
+
+        $username = $this->_config['username'];
+        $password = $this->_config['password'];
+
+        $replyCode = $this->_smtpSend(
+            sprintf(
+                'AUTH PLAIN %s',
+                base64_encode($username . chr(0) . $username . chr(0) . $password)
```
[the remark this answers is not in the document]
Comment (ADmad) [reply src/Mailer/Transport/SmtpTransport.php:284]: https://github.com/pear/Net_SMTP/blob/7e2e139628c43c40b7302136939078e19791d15a/Net/SMTP.php#L871

So the prefix is optional.

[the remark this answers is not in the document]
Comment (garas) [reply src/Mailer/Transport/SmtpTransport.php:284]: Both are used.

`\0user\0pass`

- https://github.com/laminas/laminas-mail/blob/master/src/Protocol/Smtp/Auth/Plain.php#L76
- https://github.com/PHPMailer/PHPMailer/blob/master/src/SMTP.php#L516

`user\0user\0pass`

- https://github.com/swiftmailer/swiftmailer/blob/master/lib/classes/Swift/Transport/Esmtp/Auth/PlainAuthenticator.php#L34
- https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Mailer/Transport/Smtp/Auth/PlainAuthenticator.php#L35

[the remark this answers is not in the document]
Comment (garas) [reply src/Mailer/Transport/SmtpTransport.php:284]: >    When no authorization identity is provided, the server derives an
>    authorization identity from the prepared representation of the
>    provided authentication identity string.  This ensures that the
>    derivation of different representations of the authentication
>    identity produces the same authorization identity.

So if _authzid_ is not provided, it is derived from _authcid_. So save few bytes and skip the prefix.

**Comment by markstory** [comment]
Nice work :clap:

**Pull request merged**